taichi/backends/cuda/cupti_toolkit.cpp

```diff
@@ -1,35 +1,792 @@
 #include "taichi/backends/cuda/cupti_toolkit.h"
```

Comment:
Regarding this file, which part is our own logic 😂 ?

Reply:
I was wondering if I should wrap the CUPTI code and put the `CuptiToolkit` implementation (methods `init_cupti()`, `begin_profiling()` ... ) into `cudaprofiler.cpp` like `CudaEvent` ([[link](https://github.com/taichi-dev/taichi/blob/master/taichi/backends/cuda/cuda_profiler.cpp#L161-L167)]) , so that people don’t need to look at the underlying details.

Reply:
SG, we can do it in another PR